### PR TITLE
Add atomic reindex support, via --atomic flag in search:import

### LINKED ISSUE #292: Add `search:re-import` command
The **goal** here is to add `search:re-import` command, that uses the `temporary index` technique behind the scenes.

About this command: 
```
Push a new set of objects and remove all previous ones. Settings, synonyms and query rules are untouched. Replace all records in an index without any downtime.

Internally, this method copies the existing index settings, synonyms and query rules and indexes all passed objects. Finally, the existing index is replaced by the temporary one.
```

Here is an example of this tecnique implemented on scout extented: https://github.com/algolia/scout-extended/blob/master/src/Console/Commands/ReImportCommand.php.

### PR DESCRIPTION
**Pull request by ostrolucky**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes
| BC breaks?        | no     
| Related Issue     | Fix #292
| Need Doc update   | yes

0 downtime `search:clear && search:import` solution

Screenshot 2019-11-16 at 19 22 17

### CODE REVIEW

File: src/Command/SearchImportCommand.php [hunk 1]
```diff
@@ -65,8 +90,15 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
 
         foreach ($entitiesToIndex as $entityClassName) {
-            $manager    = $this->managerRegistry->getManagerForClass($entityClassName);
-            $repository = $manager->getRepository($entityClassName);
+            $manager       = $this->managerRegistry->getManagerForClass($entityClassName);
+            $repository    = $manager->getRepository($entityClassName);
+            $realIndexName = $this->searchService->searchableAs($entityClassName);
+
+            if ($shouldDoAtomicReidex) {
+                $tmpIndex = $this->searchServiceForAtomicReindex->searchableAs($entityClassName);
```
Comment (ostrolucky) [inline src/Command/SearchImportCommand.php]: Scout-extended is doing here try {} catch {}, but command does handle situation with missing original index fine according my testing.

**Comment by ostrolucky** [comment]
Nothing I can do with test failure

**Comment by chloelbn** [comment]
Hey @ostrolucky! Thanks for your PR. As we are getting ready for the release and this feature can be delivered as a minor, I'll postpone the review a little bit. Thanks for your patience!

**Comment by ostrolucky** [comment]
Rebased on master. 4.0 is out for some time. Any chance to take a look at this now?

**Comment by chloelbn** [comment]
Hi @ostrolucky , thank you for your patience! Overall it looks good but could you tell me why you favored a `--clear` option over a whole new `reimport` command like in Scout Extended?

**Comment by ostrolucky** [comment]
Two reasons
1. Code reuse. Alternative would be to duplicate lot of same code in both commands and you would have to keep changing both places each time you update one of them, or come up with elaborate way to extract shared logic to separate classes
1. Semantics. `search:reimport` name doesn't tell people anything about its difference from `search:import`. Quite contrary, it makes people falsely assume `search:import` can be used only when index is empty. And it still doesn't say `search:reimport` is idempotent, nor that it gets rid of data in index not available in database anymore. It's also easier for users to learn new flag of `search:import` rather than whole new command, which does almost same thing. It's IMO more intuitive to do all your import business through one command.

File: src/Command/SearchImportCommand.php [hunk 2]
```diff
@@ -62,8 +87,15 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
```
Comment (chloelbn) [inline src/Command/SearchImportCommand.php]: you can just change the $tmpIndex here and add `tmp_` or any name. That way you don't have to register the `SearchService` two times

File: src/Command/SearchImportCommand.php [hunk 3]
```diff
@@ -49,8 +72,10 @@ protected function configure()
 
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        $entitiesToIndex = $this->getEntitiesFromArgs($input, $output);
-        $config          = $this->searchService->getConfiguration();
+        $shouldDoAtomicReidex = $input->getOption('clear');
```
Comment (chloelbn) [inline src/Command/SearchImportCommand.php]: `shouldDoAtomicReindex`

**Comment by chloelbn** [comment]
@ostrolucky I understand your explanation. In that case I'd rather change the flag for `--atomic`

File: src/Command/SearchImportCommand.php (hunk 2 above)
Comment (ostrolucky) [reply src/Command/SearchImportCommand.php]: I do, because of `$indexingService->index()` call

**Comment by ostrolucky** [comment]
It doesn't convey that data is imported from scratch and obsolete data will be erased. How about `--replace`?

**Comment by chloelbn** [comment]
@ostrolucky it's the `search:import` command that does this, and your feature adds the atomic dimension to it. This is why I think the `--atomic` flag is the best. WDYT?

**Comment by ostrolucky** [comment]
Indeed, failure with testSearchImportAggregator was valid. Fixed and renamed clear to atomic.

File: src/DependencyInjection/AlgoliaSearchExtension.php [hunk 4]
```diff
@@ -71,7 +73,11 @@ public function load(array $configs, ContainerBuilder $container)
                 $engineDefinition,
                 $config,
             ]
-        ))->setPublic(true);
+        ));
+
+        $searchServiceDefinitionForAtomicReindex = (clone $searchServiceDefinition)
+            ->replaceArgument(2, ['prefix' => 'atomic_reindex_in_progress_' . $config['prefix']] + $config)
```
Comment (nunomaduro) [inline src/DependencyInjection/AlgoliaSearchExtension.php]: Can we make this `atomic_reindex_in_progress_` a random name? Example:
```php
'atomic_temporary_'.uniqid('php_', true)
```

File: src/DependencyInjection/AlgoliaSearchExtension.php [hunk 5]
```diff
@@ -81,7 +87,8 @@ public function load(array $configs, ContainerBuilder $container)
             ]
         ))->setPublic(true);
 
-        $container->setDefinition('search.service', $searchServiceDefinition);
+        $container->setDefinition('search.service', $searchServiceDefinition->setPublic(true));
+        $container->setDefinition('search.service_for_atomic_reindex', $searchServiceDefinitionForAtomicReindex);
```
Comment (nunomaduro) [inline src/DependencyInjection/AlgoliaSearchExtension.php]: Just to confirm, this service is internal? People can't actually use it right?

File: src/DependencyInjection/AlgoliaSearchExtension.php (hunk 4 above)
Comment (ostrolucky) [reply src/DependencyInjection/AlgoliaSearchExtension.php]: I don't think we can do it with your current test setup, as it seems you are whitelisting index names

File: src/DependencyInjection/AlgoliaSearchExtension.php (hunk 5 above)
Comment (ostrolucky) [reply src/DependencyInjection/AlgoliaSearchExtension.php]: It's private, not internal

File: src/Command/SearchImportCommand.php [hunk 6]
```diff
@@ -62,8 +87,19 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
         foreach ($entitiesToIndex as $entityClassName) {
-            $manager    = $this->managerRegistry->getManagerForClass($entityClassName);
-            $repository = $manager->getRepository($entityClassName);
+            if (!$this->searchService->isSearchable($entityClassName)) {
+                continue;
+            }
+
+            $manager       = $this->managerRegistry->getManagerForClass($entityClassName);
+            $repository    = $manager->getRepository($entityClassName);
+            $realIndexName = $this->searchService->searchableAs($entityClassName);
```
[the remark this answers is not in the document]
Comment (ostrolucky) [reply src/Command/SearchImportCommand.php]: I don't think so, distinction between temporary and non-temporary index name must be clear.

[the remark this answers is not in the document]
Comment (nunomaduro) [reply src/Command/SearchImportCommand.php]: Maybe `destinationIndexName`?

File: src/DependencyInjection/AlgoliaSearchExtension.php (hunk 4 above)
Comment (nunomaduro) [reply src/DependencyInjection/AlgoliaSearchExtension.php]: No worries, make a temporary index name, @chloelbn will make sure it works on the test suite.

**Comment by ostrolucky** [comment]
> One very important thing tho, we should allow the user to pass a --safe, that should call the wait method on the copy, on the sum of all responses ( after the while ), and on the very last move

Is it really useful? Algolia queues up these operations. IMHO it's useful only when having following operation retrieving data from algolia, which we don't do here. See from https://www.algolia.com/doc/api-client/methods/indexing/#asynchronous-methods: "*What you are actually doing when calling these methods is adding a new job to a queue*". My testing confirms that. Can you think of some scenario to reproduce failing of this command not using wait methods?

**Comment by nunomaduro** [comment]
@ostrolucky 

> Is it really useful?

Yes, but let's implement it later in another pull request ( doesn't have to be done by you! ).

Thanks for all the work so far.

File: src/Command/SearchImportCommand.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (ostrolucky) [reply src/Command/SearchImportCommand.php]: Or sourceIndexName?

**Comment by ostrolucky** [comment]
Alright, I think I addressed everything :)

**Comment by chloelbn** [comment]
Would you mind rebasing on master please?

**Comment by ostrolucky** [comment]
done

**Pull request merged**